app/src/main/java/com/example/drawingApp/GradientSquare.kt

```diff
@@ -0,0 +1,172 @@
+                }
+                MotionEvent.ACTION_MOVE -> {
+                    if (held) {
+                        onSatOrValChange(sat, valu)
+                    }
+                }
+            }
+            true
+        }
+    }
+
+    private fun startCacheTask() {
+        GlobalScope.launch(Dispatchers.IO) {
+            regenerate()
+            completedIndexes?.let {
+                val startPoint = completedIndexes
+                if (startPoint != null) {
+                    for (i in startPoint..359) {
+                        regenerate()
+                        bitmaps?.set(i, generateBitmap(i))
+                        completedIndexes = completedIndexes?.plus(1)
+                    }
+                    areBitmapsDone = true
+                    isCaching = false
+                }
+            }
+        }
+    }
+
+    private fun getBitmap(): Bitmap? {
+        val rIndex = max(0, min(hue.toInt(), 359))
+        regenerate()
+        return if (bitmaps?.get(rIndex) == null) {
+            bitmaps?.set(rIndex, generateBitmap(hue.toInt()))
+            bitmaps?.get(rIndex)
+        } else bitmaps?.get(rIndex)
+    }
+
+    //needs the selectedHue as we are creating the array of bitmaps in the background by passing it the int
+    private fun generateBitmap(selectedHue: Int): Bitmap {
+        val hStep = 1f / imageSize
+        val vStep = 1f / imageSize
+        var v = 1f
+        var s = 0f
+        val hsv = floatArrayOf(selectedHue.toFloat(), 1f, 1f)
+
+        var index = 0
+        val createdBitmap = Bitmap.createBitmap(imageSize, imageSize, Bitmap.Config.ARGB_8888)
+        val pixels = IntArray(imageSize * imageSize)
+
+        repeat(imageSize) {
+            repeat(imageSize) {
+                hsv[1] = s
+                hsv[2] = v
+                pixels[index] = Color.HSVToColor(hsv)
+                s += hStep
+                index += 1
+            }
+            s = 0f
+            v -= vStep
+        }
+        createdBitmap.setPixels(pixels, 0, imageSize, 0, 0, imageSize, imageSize)
+        return createdBitmap
+    }
+
+    @SuppressLint("DrawAllocation")
+    override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+        canvas?.let { c ->
+            if (!::drawRect.isInitialized) drawRect = Rect(0, 0, width, height)
+            synchronized(lock) {
+                if (bitmaps == null || isCaching == null) regenerate()
+                areBitmapsDone?.let { bits ->
+                    isCaching?.let { cache ->
+                        if (!cache && !bits) {
+                            isCaching = true
+                            startCacheTask()
+                        }
+                    }
+                }
+            }
+            getBitmap()?.let { c.drawBitmap(it, null, drawRect, null) }
+            c.drawCircle(sat * width, height - valu * height, 5f, selectionPaint)
+        }
+    }
+
+    companion object {
+        const val imageSize = 175
```

Comment:
we can go SMALLER

Reply:
we can, but I started noticing visual differences at 150 and smaller looks worse.

Reply:
175 is already really fast